File: Regression/CppR/driver.cpp
//*********************************************************
// This file is part of Rapsodia released under the LGPL. *
// The full COPYRIGHT notice can be found in the top      *
// level directory of the Rapsodia distribution           *
//*********************************************************
#include <iostream>
#include <cmath>

#include "RAinclude.ipp"


int main(void){
  int rc=0;
  const int directions=2;
  const int order=10;

  int testCase;
  bool done, dumpedTestLine;
  double xCoeff[directions][order],yCoeff[directions][order];
  RAfloatD  x,y;
  double xp,yp, myEps=1.E-13;
  const char* testLine;
  int i,j;
  
  // the point at which we test
  xp=0.3;
  
  // keep the coefficient values for comparison
  for( i=0;i<directions;i++) { 
    for( j=0;j<order; j++) { 
      xCoeff[i][j]=0.0;
    }
    xCoeff[i][1]=i*0.3;
  }
 
  // initialize the active input
  x=xp;
  for( i=0;i<directions;i++) { 
    for( j=0;j<order; j++) { 
      x.set(i+1,j+1,xCoeff[i][j]);
    }
  } 

  // populate the coefficients
  // with something nontrivial
  x=asin(x);

  done=false;
  testCase=0;

  
  // run the identities
  while (!done) { 
    switch(testCase) { 
    case 0:
      y=x;
      testLine="y=x";
      break;
    case 1: 
      y=x+x-x;
      testLine="y=x+x-x";
      break;
    case 2:
      y=-2+(2+((x+2)-2));
      testLine="y=-2+(2+((x+2)-2))";
      break;
    case 3:
      y=(2*x)/2;
      testLine="y=(2*x)/2";
      break;
    case 4:
      y=(x*x)/x;
      testLine="y=(x*x)/x";
      break;
    case 5:
      y=sqrt(x*x);
      testLine="y=sqrt(x*x)";
      break;
    case 6:
      y=sqrt(pow(x,2));
      testLine="y=sqrt(pow(x,2))";
      break;
    case 7:
      y=log(exp(x));
      testLine="y=log(exp(x))";
      break;
    case 8:
      y=pow(pow(x,2),.5);
      testLine="y=pow(pow(x,2),.5)";
      break;
    case 9:
      y=pow(pow(x,3),1.0/3.0);
      testLine="y=pow(pow(x,3),1.0/3.0)";
      break;
    case 10:
      y=atan(tan(x));
      testLine="y=atan(tan(x))";
      break;
    case 11:
      y=(-x)*(-1.0);
      testLine="y=(-x)*(-1.0);";
      break;
    case 12:
      y=0.0;
      y+=2*x;
      y-=x;
      testLine="y+=2*x;y-=x";
      break;
    case 13:
      y=0.0;
      y+=x;
      y*=x;
      y/=x;
      testLine="y+=x;y*=x;y/=x";
      break;
    default:
      done=true;
      break;
    }
    if (!done) { 
      y=sin(y);
      dumpedTestLine=false;
      // test the value: 
      if (fabs(y.v-xp)>myEps) { 
	if (!dumpedTestLine) {
	  std::cout << "deviation for " <<  testLine << std::endl; 
	  dumpedTestLine=true;
	}
	std::cout << " y.v=" << y.v << " x.v=" << xp << " diff: " <<  fabs(y.v-xp) << std::endl;
        rc=1; 
      }
      // test the coefficients
      for( i=0;i<directions;i++) { 
	for( j=0;j<order; j++) { 
	  yCoeff[i][j]=y.get(i+1,j+1);
	  if (fabs(yCoeff[i][j]-xCoeff[i][j])>myEps) { 
	    if (!dumpedTestLine) {
	      std::cout << "deviation for " <<  testLine << std::endl; 
	      dumpedTestLine=true;
	    }
	    std::cout << " y[" << i << "][" << j << "]=" << yCoeff[i][j] << " x[" << i << "][" << j << "]=" << xCoeff[i][j] << " diff: " << fabs(yCoeff[i][j]-xCoeff[i][j]) << std::endl; 
            rc=1;
	  }
	}
      }
      if (!dumpedTestLine) {
	std::cout << "test " << testLine << " ok" << std::endl; 
      }
    }
    testCase=testCase+1;
  }
  return rc;
}

File: Regression/F90R/driver.f90
!#########################################################
! This file is part of Rapsodia released under the LGPL. #
! The full COPYRIGHT notice can be found in the top      #
! level directory of the Rapsodia distribution           #
!#########################################################
 program identities

  include 'RAinclude.i90'
  
  integer, parameter :: directions=2
  integer, parameter :: order=10

  integer testCase
  logical done, dumpedTestLine
  real(kind=RAdKind), dimension(directions,order) :: xCoeff,yCoeff
  type(RArealD) :: x,y
  real(kind=RAdKind) :: xp,yp, myEps=1.E-12
  character(len=50) :: testLine
  integer i,j

  ! the point at which we test
  xp=0.3
  
  ! keep the coefficient values for comparison
  do i=1,directions
    do j=1,order
      xCoeff(i,j)=0.0
    end do
    xCoeff(i,1)=REAL(i,kind=RAdKind)*0.3D0
  end do
 
  ! initialize the active input
  x=xp
  do i=1,directions
    do j=1,order
       call RAset(x,i,j,xCoeff(i,j))
    end do
  end do
 
  ! populate the coefficients
  ! with something nontrivial
  x=asin(x)

  done=.false.
  testCase=0

  ! run the identities
  do while (.not.done) 

   select case(testCase)
  
     case(0) 
        y=x
        ! print *, x
        ! print *, y
        testline='y=x'

     case(1) 
        y=x+x-x
        testline='y=x+x-x'

     case(2)
        y=-2+(2+((x+2)-2))
        testline='y=-2+(2+((x+2)-2))'

     case(3)
        y=(2*x)/2
        testline='y=(2*x)/2'

     case(4)
        y=(x*x)/x
        testline='y=(x*x)/x'

     case(5)
        y=sqrt(x*x)
        testline='y=sqrt(x*x)'

     case(6)
        y=sqrt(x**2)
        testline='y=sqrt(x**2)'

     case(7)
        y=log(exp(x))
        testline='y=log(exp(x))'

     case(8)
        y=(x**2)**(.5)
        testline='y=(x**2)**(.5)'

     case(9)
        y=(x**3)**(1.0D0/3.0D0)
        testline='y=(x**3)**(1.0D0/3.0D0)'

     case(10)
        y=atan(tan(x))
        testline='y=atan(tan(x))'

     case(11)
        y=tan(atan(x))
        testline='y=tan(atan(x))'

     case(12)
        y=(-x)*(-1.0D0)
        testline='y=(-x)*(-1.0D0)'

     case default
        done=.true.

  end select
  if (.not.done) then 
     y=sin(y)
     dumpedTestLine=.false.
     ! test the value: 
     if (abs(y%v-xp)>myEps) then
       if (.not.dumpedTestLine) then 
         write(*,'(A,A)') 'deviation for ', testLine
         dumpedTestLine=.true.
       end if
       write(*,'(A,E25.17E3,A,E25.17E3,A,E22.14E3)')  '  y%v=',y%v,' x%v=',xp, ' diff: ', abs(y%v-xp)
       stop 1
     end if 
     ! test the coefficients
     do i=1,directions
       do j=1,order
        call RAget(y,i,j,yCoeff(i,j))
        if (abs(yCoeff(i,j)-xCoeff(i,j))>myEps) then
          if (.not.dumpedTestLine) then 
            write(*,'(A,A)') 'deviation for ', testLine
            dumpedTestLine=.true.
          end if
          write(*,'(A,I1,A,I2,A,E25.17E3,A,I1,A,I2,A,E25.17E3,A,E25.17E3)') '  y(',i,',',j,')=',&
yCoeff(i,j),' x(',i,',',j,')=',xCoeff(i,j), ' diff: ', abs(yCoeff(i,j)-xCoeff(i,j))
          stop 1
        end if
       end do
     end do
     if (.not.dumpedTestLine) then 
       write(*,'(A,A,A)') 'test ', testLine, ' ok'
     end if            
  end if
  testCase=testCase+1
  end do

end program 
